src/main/java/hudson/plugins/ec2/EC2Cloud.java

```diff
@@ -751,7 +793,11 @@ private static AmazonWebServicesCredentials getCredentials(@Nullable String cred
 
     private AmazonEC2 reconnectToEc2() throws IOException {
         synchronized(this) {
-            connection = connect(createCredentialsProvider(), getEc2EndpointUrl());
+            if (AmazonEC2Cloud.isTestMode()) {
```

Comment:
I am not maintainer or owner of this plugin. Just happen to see this and adding my review comments.

I am not a big fan of mixing mock code in original code. Some suggestions here https://8thlight.com/blog/dariusz-pasciak/2014/12/19/dont-mix-test-code-with-production-code.html

Reply:
Well, no, generally you don't want that, but as you can see in the code this methodology is already used, and frankly I'm not sure Jenkins supports injecting and/or replacing (test) classes in a running Jenkins instance.

That blog post deals mostly with unit testing.

Reply:
We should clean up this test mode stuff, but I'd be happy seeing it done in a separate PR since it's both pervasive already in this plugin, and fixing it here would make the diff gigantic.